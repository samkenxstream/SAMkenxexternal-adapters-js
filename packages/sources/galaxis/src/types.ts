import { ethers } from 'ethers'

export interface FilteredAchievements {
  encodedCalls: string
  hasMore: boolean
}

export interface Achievement {
  team_id?: number
  player_id?: number
  achievement_id: number
  value: boolean | number
}

export interface TeamStruct {
  id: number
  name: string
  city: string
  tricode: string
  real_id: ethers.BigNumber
}

export interface PlayerStruct {
  id: number
  team_id: number
  real_id: ethers.BigNumber
  real_team_id: ethers.BigNumber
  full_name: string
}

export interface AchievementsByIDs {
  [T: string]: Achievement[]
}
