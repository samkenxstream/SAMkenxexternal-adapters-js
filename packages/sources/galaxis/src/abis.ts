export const EC_REGISTRY_ABI = [
  {
    inputs: [
      { internalType: 'uint16', name: 'traitId', type: 'uint16' },
      { internalType: 'uint16[]', name: '_ids', type: 'uint16[]' },
      { internalType: 'uint8[]', name: '_data', type: 'uint8[]' },
    ],
    name: 'setData',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    inputs: [
      { internalType: 'address', name: '_addr', type: 'address' },
      { internalType: 'uint16', name: 'traitID', type: 'uint16' },
    ],
    name: 'addressCanModifyTrait',
    outputs: [{ internalType: 'bool', name: 'result', type: 'bool' }],
    stateMutability: 'view',
    type: 'function',
  },
  {
    inputs: [{ internalType: 'uint16', name: '', type: 'uint16' }],
    name: 'traits',
    outputs: [
      { internalType: 'uint16', name: 'id', type: 'uint16' },
      { internalType: 'uint8', name: 'traitType', type: 'uint8' },
      { internalType: 'uint16', name: 'start', type: 'uint16' },
      { internalType: 'uint16', name: 'end', type: 'uint16' },
      { internalType: 'address', name: 'implementer', type: 'address' },
      { internalType: 'string', name: 'name', type: 'string' },
    ],
    stateMutability: 'view',
    type: 'function',
  },
]

export const EC_REGISTRY_MAP_ABI = [
  {
    inputs: [],
    name: 'getPlayers',
    outputs: [
      {
        components: [
          { internalType: 'uint16', name: 'id', type: 'uint16' },
          { internalType: 'uint16', name: 'team_id', type: 'uint16' },
          { internalType: 'uint256', name: 'real_id', type: 'uint256' },
          { internalType: 'uint256', name: 'real_team_id', type: 'uint256' },
          { internalType: 'string', name: 'full_name', type: 'string' },
        ],
        internalType: 'struct ECRegistryMap.playerStruct[]',
        name: '',
        type: 'tuple[]',
      },
    ],
    stateMutability: 'view',
    type: 'function',
  },
  {
    inputs: [],
    name: 'getTeams',
    outputs: [
      {
        components: [
          { internalType: 'uint16', name: 'id', type: 'uint16' },
          { internalType: 'string', name: 'name', type: 'string' },
          { internalType: 'string', name: 'city', type: 'string' },
          { internalType: 'string', name: 'tricode', type: 'string' },
          { internalType: 'uint256', name: 'real_id', type: 'uint256' },
        ],
        internalType: 'struct ECRegistryMap.teamStruct[]',
        name: '',
        type: 'tuple[]',
      },
    ],
    stateMutability: 'view',
    type: 'function',
  },
]

export const TRAIT_IMPLEMENTER_ABI = [
  {
    inputs: [
      { internalType: 'uint16[]', name: '_tokenIds', type: 'uint16[]' },
      { internalType: 'uint8[]', name: '_value', type: 'uint8[]' },
    ],
    name: 'setData',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
]

export const BATCH_WRITER_ABI = [
  {
    inputs: [{ internalType: 'uint16', name: 'id', type: 'uint16' }],
    name: 'CallFailed',
    type: 'error',
  },
  {
    anonymous: false,
    inputs: [
      { indexed: true, internalType: 'address', name: 'previousOwner', type: 'address' },
      { indexed: true, internalType: 'address', name: 'newOwner', type: 'address' },
    ],
    name: 'OwnershipTransferred',
    type: 'event',
  },
  {
    anonymous: false,
    inputs: [
      { indexed: true, internalType: 'bytes32', name: 'requestId', type: 'bytes32' },
      { indexed: true, internalType: 'bytes', name: 'data', type: 'bytes' },
    ],
    name: 'RequestFulfilled',
    type: 'event',
  },
  {
    inputs: [{ internalType: 'address', name: '_addr', type: 'address' }],
    name: 'canCall',
    outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
    stateMutability: 'view',
    type: 'function',
  },
  {
    inputs: [
      { internalType: 'bytes32', name: 'requestId', type: 'bytes32' },
      { internalType: 'bytes', name: 'bytesData', type: 'bytes' },
    ],
    name: 'fulfillBytes',
    outputs: [{ internalType: 'bool[]', name: '', type: 'bool[]' }],
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    inputs: [],
    name: 'owner',
    outputs: [{ internalType: 'address', name: '', type: 'address' }],
    stateMutability: 'view',
    type: 'function',
  },
  {
    inputs: [],
    name: 'renounceOwnership',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    inputs: [],
    name: 'requestBytes',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    inputs: [{ internalType: 'address', name: 'newOwner', type: 'address' }],
    name: 'transferOwnership',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
]
