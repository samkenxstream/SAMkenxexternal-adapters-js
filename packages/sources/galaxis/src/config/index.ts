import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'GALAXIS'

export const DEFAULT_ENDPOINT = 'nba'
export const DEFAULT_BASE_URL = 'http://localhost:18081'
export const DEFAULT_EC_REGISTRY_ADDRESS = '0x163883263274e8Ef6332cFa84F35B23c6C51dF72'
export const DEFAULT_BATCH_WRITER_ADDRESS = '0xB57fba975C89492B016e0215E819B4d489F0fbcD'
export const DEFAULT_LIMIT_IN_BYTES = 250

export interface ExtendedConfig extends Config {
  ecRegistryAddress: string
  batchWriterAddress: string
  maxEncodedCallsBytes: number
}

export const makeConfig = (prefix?: string): ExtendedConfig => {
  const config = Requester.getDefaultConfig(prefix)
  config.api.baseURL = config.api.baseURL || DEFAULT_BASE_URL
  config.defaultEndpoint = DEFAULT_ENDPOINT
  const maxEncodedCallsBytes = util.getEnv('LIMIT', prefix)
  return {
    ...config,
    ecRegistryAddress: util.getEnv('EC_REGISTRY_ADDRESS', prefix) || DEFAULT_EC_REGISTRY_ADDRESS,
    batchWriterAddress: util.getEnv('BATCH_WRITER_ADDRESS', prefix) || DEFAULT_BATCH_WRITER_ADDRESS,
    maxEncodedCallsBytes: maxEncodedCallsBytes
      ? parseInt(maxEncodedCallsBytes)
      : DEFAULT_LIMIT_IN_BYTES,
  }
}
