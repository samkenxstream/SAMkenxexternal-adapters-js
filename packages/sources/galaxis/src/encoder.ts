import { ByteArray } from '@ethercards/ec-util'

export const callToRequestData = (calls: string[][], dataRecordId: number): string => {
  let bytes = ''
  const header = new ByteArray(Buffer.alloc(2))
  // add call num
  header.writeUnsignedShort(calls.length)
  // add record id
  header.writeUnsignedInt(dataRecordId)
  bytes = header.toString('hex')

  for (let i = 0; i < calls.length; i++) {
    const callLen = callLentoHex(removeZeroX(calls[i][1]).length)
    const address = addresstoCallData(calls[i][0])
    const callData = removeZeroX(calls[i][1])
    const packet = callLen + address + callData
    bytes += packet
  }
  return bytes
}

const removeZeroX = (str: string): string => {
  return str.replace('0x', '')
}

const addresstoCallData = (str: string): string => {
  return '000000000000000000000000' + removeZeroX(str)
}

const callLentoHex = (num: number): string => {
  const data = new ByteArray(Buffer.alloc(2))
  data.writeUnsignedShort(num / 2)
  return removeZeroX(data.toString('hex'))
}
