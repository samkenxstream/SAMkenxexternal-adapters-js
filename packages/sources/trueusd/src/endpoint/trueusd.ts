import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { ExecuteWithConfig, Config, InputParameters } from '@chainlink/ea-bootstrap'

export const NAME = 'trueusd'

export const supportedEndpoints = ['trueusd']

export const endpointResultPaths = {
  trueusd: 'totalTrust',
}

const customError = (data: ResponseSchema) => !data.success

export const description = 'https://core-api.real-time-attest.trustexplorer.io/trusttoken/TrueUSD'

export type TInputParameters = Record<string, never>
export const inputParameters: InputParameters<TInputParameters> = {}

interface ResponseSchema {
  responseData: {
    accountName: string
    totalTrust: number
    totalToken: number
    updatedAt: string
    token: { tokenName: string; principle: number }[]
  }
  message: { msg: string }[]
  success: boolean
  responseCode: number
}

export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
  const validator = new Validator<TInputParameters>(request, inputParameters)

  const jobRunID = validator.validated.id
  const resultPath = (validator.validated.data.resultPath || '').toString()
  const url = '/trusttoken/TrueUSD'

  const options = { ...config.api, url }

  const response = await Requester.request<ResponseSchema>(options, customError)

  const result = Requester.validateResultNumber(response.data, ['responseData', resultPath])

  return Requester.success(jobRunID, Requester.withResult(response, result), config.verbose)
}
