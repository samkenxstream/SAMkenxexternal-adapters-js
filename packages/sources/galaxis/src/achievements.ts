import { ethers } from 'ethers'
import {
  BATCH_WRITER_ABI,
  EC_REGISTRY_ABI,
  EC_REGISTRY_MAP_ABI,
  TRAIT_IMPLEMENTER_ABI,
} from './abis'
import { callToRequestData } from './encoder'
import { ExtendedConfig } from './config'
import {
  Achievement,
  AchievementsByIDs,
  AchievementWithMappedID,
  PlayerStruct,
  TeamStruct,
} from './types'
import { AdapterError, Logger } from '@chainlink/ea-bootstrap'

export const getEncodedCallsResult = async (
  jobRunID: string,
  playerAchievements: Achievement[],
  teamAchievements: Achievement[],
  config: ExtendedConfig,
): Promise<string> => {
  const groupedAchievements = getGroupedAchievementsByID(playerAchievements, teamAchievements)
  const provider = new ethers.providers.JsonRpcProvider(config.rpcUrl)
  const { teams, players } = await getPlayerAndTeamMaps(provider, config)
  const calls: string[][] = []

  let hasHitLimit = false
  let encodedCalls = ''
  const achievementIDs = Object.keys(groupedAchievements).sort((a, b) => parseInt(a) - parseInt(b))
  const batchWriter = new ethers.Contract(config.batchWriterAddress, BATCH_WRITER_ABI, provider)
  const ecRegistry = new ethers.Contract(config.ecRegistryAddress, EC_REGISTRY_ABI, provider)

  const lastProcessedInfo = await getIdxLastProcessedAchievementID(achievementIDs, batchWriter)
  let currAchievementIdIdx = lastProcessedInfo.idx
  while (!hasHitLimit && currAchievementIdIdx < achievementIDs.length) {
    const achievementID = achievementIDs[currAchievementIdIdx]
    if (await ecRegistry.addressCanModifyTrait(config.batchWriterAddress, achievementID)) {
      const encodedCall = await getSetDataEncodedCall(
        provider,
        teams,
        players,
        groupedAchievements,
        achievementID,
        ecRegistry,
      )
      calls.push([config.ecRegistryAddress, encodedCall])
      const { encodedCalls: updatedEncodedCalls, hasHitLimit: hasCallsHitLimit } =
        await updateEncodedCalls(jobRunID, achievementID, calls, batchWriter)
      hasHitLimit = hasCallsHitLimit
      if (!hasHitLimit) {
        encodedCalls = updatedEncodedCalls
      }
    }
    currAchievementIdIdx++
  }
  const lastProcessedID = lastProcessedInfo.lastProcessedID
  validateEncodedCallsNotEmpty(jobRunID, lastProcessedID, encodedCalls)
  return encodedCalls
}

const validateEncodedCallsNotEmpty = (
  jobRunID: string,
  lastProcessedID: string,
  encodedCalls: string,
) => {
  if (encodedCalls.length === 0) {
    throw new AdapterError({
      jobRunID,
      statusCode: 500,
      message: `Got empty encoded results when resuming from achievementID ${lastProcessedID}`,
    })
  }
}

const updateEncodedCalls = async (
  jobRunID: string,
  achievementID: string,
  calls: string[][],
  batchWriter: ethers.Contract,
): Promise<{
  hasHitLimit: boolean
  encodedCalls: string
}> => {
  const encodedCalls = callToRequestData(calls, parseInt(achievementID))
  let hasHitLimit = false
  try {
    const gasCostEstimate = await batchWriter.estimateGas.fulfillBytes(
      ethers.utils.formatBytes32String(jobRunID),
      `0x${encodedCalls}`,
    )
    Logger.info(
      `Successfully estimated gas ${gasCostEstimate.toString()} for processing achievementID ${achievementID}`,
    )
  } catch (e) {
    if (e.code === 'UNPREDICTABLE_GAS_LIMIT') {
      hasHitLimit = true
      Logger.info(`Hit gas limit when processing achievementID ${achievementID}`)
    } else {
      throw new AdapterError({
        jobRunID,
        message: e.message,
        statusCode: 500,
      })
    }
  }
  return {
    encodedCalls,
    hasHitLimit,
  }
}

const getIdxLastProcessedAchievementID = async (
  sortedAchievementIDs: string[],
  batchWriter: ethers.Contract,
): Promise<{
  idx: number
  lastProcessedID: string
}> => {
  const lastProcessedID = await batchWriter.LastDataRecordId()
  const lastProcessedIDIdx = sortedAchievementIDs.findIndex(
    (achievementID) => achievementID === lastProcessedID.toString(),
  )
  if (lastProcessedIDIdx < 0) {
    Logger.info(`Cannot find achievementID ${lastProcessedID}.  Will process all achievements`)
    return {
      idx: 0,
      lastProcessedID: lastProcessedID.toString(),
    }
  }
  return {
    idx: lastProcessedIDIdx,
    lastProcessedID: lastProcessedID.toString(),
  }
}

const getSetDataEncodedCall = async (
  provider: ethers.providers.JsonRpcProvider,
  teams: TeamStruct[],
  players: PlayerStruct[],
  groupedAchievements: AchievementsByIDs,
  achievementID: string,
  ecRegistry: ethers.Contract,
): Promise<string> => {
  const achievements = getAchievementsWithMappedIDs(
    teams,
    players,
    groupedAchievements,
    achievementID,
  )
  if (isOnlyBooleanValues(groupedAchievements[achievementID].map(({ value }) => value))) {
    return getSetDataEncodedCallForOnlyBooleanAchievement(ecRegistry, achievements, achievementID)
  }
  return await getSetDataEncodedCallForMixedValueAchievement(
    provider,
    ecRegistry,
    achievements,
    achievementID,
  )
}

const getAchievementsWithMappedIDs = (
  teams: TeamStruct[],
  players: PlayerStruct[],
  groupedAchievements: AchievementsByIDs,
  achievementID: string,
): AchievementWithMappedID[] => {
  return groupedAchievements[achievementID].map(({ team_id, player_id, ...rest }) => {
    let mappedID
    if (team_id) {
      const team = teams.find((t) => t.real_id.toString() === team_id.toString())
      if (!team) {
        throw new Error(`Cannot match team ID ${team_id} with response from ECRegistryMap`)
      }
      mappedID = team.id
    } else if (player_id) {
      const player = players.find((p) => p.real_id.toString() === player_id.toString())
      if (!player) {
        throw new Error(`Cannot match player ID ${team_id} with response from ECRegistryMap`)
      }
      mappedID = player.id
    } else {
      throw new Error('Missing player and team IDs')
    }
    return {
      ...rest,
      mappedID: mappedID,
    }
  })
}

const isOnlyBooleanValues = (values: (number | boolean)[]): boolean => {
  for (const value of values) {
    if (typeof value !== 'boolean') return false
  }
  return true
}

// Data in ECRegistry is stored in a BitMap which means we need
// to load existing data, check for differences and only update the
// new key uint8's that have been changed
// https://github.com/ethercards/ec-chain-batch-execute/blob/master/tests/batchWriter.ts#L228
const getSetDataEncodedCallForOnlyBooleanAchievement = async (
  ecRegistry: ethers.Contract,
  updatedAchievements: AchievementWithMappedID[],
  achievementID: string,
): Promise<string> => {
  const pageNumber = 0
  const numPageRecords = updatedAchievements.length
  const oldValues = await ecRegistry.getData(achievementID, pageNumber, numPageRecords)
  const valueDifferences: number[] = []
  const valueDifferenceIDs: number[] = []
  for (let z = 0; z < updatedAchievements.length; z++) {
    const boolAsNum = typeof updatedAchievements[z].value === 'boolean' ? 1 : 0
    if (oldValues[z] !== boolAsNum) {
      valueDifferenceIDs.push(updatedAchievements[z].mappedID)
      valueDifferences.push(boolAsNum)
    }
  }
  return ecRegistry.interface.encodeFunctionData('setData', [
    parseInt(achievementID),
    valueDifferenceIDs,
    valueDifferences,
  ])
}

const getSetDataEncodedCallForMixedValueAchievement = async (
  provider: ethers.providers.JsonRpcProvider,
  ecRegistry: ethers.Contract,
  achievements: AchievementWithMappedID[],
  achievementID: string,
): Promise<string> => {
  const { implementer: implementerAddress } = await ecRegistry.traits(achievementID)
  const implementer = new ethers.Contract(implementerAddress, TRAIT_IMPLEMENTER_ABI, provider)
  return implementer.interface.encodeFunctionData('setData', [
    achievements.map((a) => a.mappedID),
    achievements.map(({ value }) => (typeof value === 'boolean' ? (value ? 1 : 0) : value)),
  ])
}

const getGroupedAchievementsByID = (
  playerAchievements: Achievement[],
  teamAchievements: Achievement[],
): AchievementsByIDs => {
  let result: AchievementsByIDs = {}
  result = groupAchievements(playerAchievements, result)
  return groupAchievements(teamAchievements, result)
}

const groupAchievements = (
  achievements: Achievement[],
  achievementsByID: AchievementsByIDs,
): AchievementsByIDs => {
  return achievements.reduce((acc, curr) => {
    const { achievement_id } = curr
    if (!acc[achievement_id]) {
      acc[achievement_id] = []
    }
    acc[achievement_id].push(curr)
    return acc
  }, achievementsByID)
}

const getPlayerAndTeamMaps = async (
  provider: ethers.providers.JsonRpcProvider,
  config: ExtendedConfig,
): Promise<{ teams: TeamStruct[]; players: PlayerStruct[] }> => {
  const ecRegistryMap = new ethers.Contract(
    config.ecRegistryMapAddress,
    EC_REGISTRY_MAP_ABI,
    provider,
  )
  const teams = await ecRegistryMap.getTeams()
  const players = await ecRegistryMap.getPlayers()
  return {
    teams,
    players,
  }
}
