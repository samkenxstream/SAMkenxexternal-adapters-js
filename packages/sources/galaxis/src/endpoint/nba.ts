import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { ExecuteWithConfig, InputParameters } from '@chainlink/types'
import { ExtendedConfig } from '../config'
import { Achievement } from '../types'
import { getEncodedCallsResult } from '../achievements'

export const supportedEndpoints = ['nba']

export interface ResponseSchema {
  player_achievements: Achievement[]
  team_achievements: Achievement[]
}

const customError = (data: Record<string, unknown>) => data.Response === 'Error'

export const description =
  'This endpoint fetches a list of achievements for NBA teams and players and returns them as an encoded value'

export const inputParameters: InputParameters = {}

export const execute: ExecuteWithConfig<ExtendedConfig> = async (request, _, config) => {
  const validator = new Validator(request)
  const jobRunID = validator.validated.id
  const options = config.api
  const response = await Requester.request<ResponseSchema>(options, customError)
  const encodedCalls = await getEncodedCallsResult(
    jobRunID,
    response.data.player_achievements,
    response.data.team_achievements,
    config,
  )
  return {
    jobRunID,
    result: encodedCalls,
    statusCode: 200,
    data: {
      result: encodedCalls,
    },
  }
}
