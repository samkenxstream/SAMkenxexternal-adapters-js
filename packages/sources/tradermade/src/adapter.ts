import {
  Config,
  ExecuteWithConfig,
  MakeWSHandler,
  AdapterRequest,
  ExecuteFactory,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { Requester, Validator, Builder } from '@chainlink/ea-bootstrap'
import { makeConfig, DEFAULT_WS_API_ENDPOINT } from './config'
import * as endpoints from './endpoint'
import overrides from './config/symbols.json'

export const execute: ExecuteWithConfig<Config, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<Config, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<Config, endpoints.TInputParameters> =>
  Builder.selectEndpoint<Config, endpoints.TInputParameters>(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<Config, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

// interface Message {
//   symbol: string
//   ts: string
//   bid: number
//   ask: number
//   mid: number
// }

export const makeWSHandler = (config?: Config): MakeWSHandler | undefined => {
  const getSubscription = (pair?: string) => {
    const defaultConfig = config || makeConfig()
    if (!pair) return ''
    const sub = {
      userKey: defaultConfig.wsApiKey,
      symbol: pair,
    }
    return sub
  }
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.forex.inputParameters,
      {},
      { shouldThrowError: false, overrides },
    )
    if (validator.error) return
    const base = validator.validated.data.base.toUpperCase()
    const quote = validator.validated.data.quote.toUpperCase()
    return `${base}${quote}`
  }
  return () => {
    const defaultConfig = config || makeConfig()
    return {
      connection: {
        url: defaultConfig.ws?.baseWsURL || DEFAULT_WS_API_ENDPOINT,
      },
      shouldNotServeInputUsingWS: (input: any) =>
        endpoints.forex.supportedEndpoints.indexOf(input.data.endpoint) === -1,
      subscribe: (input: any) => getSubscription(getPair(input)),
      unsubscribe: () => undefined, // Tradermade does not support unsubscribing.
      subsFromMessage: (message: any) => {
        if (!message.symbol) return ''
        return getSubscription(message.symbol)
      },
      isError: () => false, // No error
      filter: (message: any) => !!message.mid,
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['mid'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
