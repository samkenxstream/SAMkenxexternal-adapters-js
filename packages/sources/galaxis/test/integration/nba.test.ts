import { AdapterRequest } from '@chainlink/types'
import { server as startServer } from '../../src'
import nock from 'nock'
import http from 'http'
import request, { SuperTest, Test } from 'supertest'

import { ethers } from 'ethers'
import { AddressInfo } from 'net'
import { mockGalaxisApiResp } from './fixtures'
import { Interface } from 'ethers/lib/utils'
import { EC_REGISTRY_ABI } from '../../src/endpoint/abis'

const fakeApiEndpoint = 'http://galaxis.com/achievements/'

const mockECRegistryInterface = new Interface(EC_REGISTRY_ABI)

jest.mock('ethers', () => {
  const actualModule = jest.requireActual('ethers')
  return {
    ...actualModule,
    ethers: {
      ...actualModule.ethers,
      providers: {
        JsonRpcProvider: function (_: string): ethers.provider.JsonRpcProvider {
          return {}
        },
      },
      Contract: function () {
        return {
          interface: mockECRegistryInterface,
          addressCanModifyTrait: (_, achievementID: string): boolean => {
            return achievementID === '2' || achievementID === '15'
          },
          getTeams: () => {
            return [
              {
                id: 1,
                name: 'team-1',
                city: 'Seattle',
                tricode: 'SEA',
                real_id: 1610612737,
              },
              {
                id: 2,
                name: 'team-1',
                city: 'Seattle',
                tricode: 'SEA',
                real_id: 1610612738,
              },
            ]
          },
          getPlayers: () => {
            return [
              {
                id: 3,
                team_id: 1,
                real_id: 203991,
                real_team_id: 1610612738,
                name: 'Lebron James',
              },
              {
                id: 4,
                team_id: 1,
                real_id: 203991,
                real_team_id: 1610612738,
                name: 'Steph Curry',
              },
            ]
          },
        }
      },
    },
  }
})

let oldEnv: NodeJS.ProcessEnv

let server: http.Server
let req: SuperTest<Test>
const jobID = '1'

describe('nba', () => {
  beforeEach(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    server = await startServer()
    req = request(`localhost:${(server.address() as AddressInfo).port}`)
    process.env.POLYGON_RPC_URL = process.env.POLYGON_RPC_URL || 'fake-polygon-rpc-url'
    process.env.API_ENDPOINT = process.env.API_ENDPOINT || fakeApiEndpoint
    process.env.CACHE_ENABLED = 'false'

    if (process.env.RECORD) {
      nock.recorder.rec()
    }
  })

  afterEach((done) => {
    process.env = oldEnv
    server.close(done)
  })

  describe('successful calls', () => {
    it('return success when fetching the achievements and returns the correct result', async () => {
      mockGalaxisApiResp(fakeApiEndpoint)
      const data: AdapterRequest = {
        id: jobID,
        data: {},
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
