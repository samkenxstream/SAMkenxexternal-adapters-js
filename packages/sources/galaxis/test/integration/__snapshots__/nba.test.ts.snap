// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nba successful calls return success when fetching the achievements and returns the correct result 1`] = `
Object {
  "data": Object {
    "result": "000200e4000000000000000000000000163883263274e8Ef6332cFa84F35B23c6C51dF7266a57e970000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100e4000000000000000000000000163883263274e8Ef6332cFa84F35B23c6C51dF7266a57e97000000000000000000000000000000000000000000000000000000000000000f000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000500",
  },
  "jobRunID": "1",
  "result": "000200e4000000000000000000000000163883263274e8Ef6332cFa84F35B23c6C51dF7266a57e970000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100e4000000000000000000000000163883263274e8Ef6332cFa84F35B23c6C51dF7266a57e97000000000000000000000000000000000000000000000000000000000000000f000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000500",
  "statusCode": 200,
}
`;
